include/crm/cluster/internal.h

```diff
@@ -47,6 +47,54 @@ enum pcmk__node_search_flags {
     pcmk__node_search_cluster_cib       = (1 << 2),
 };
 
+//! Cluster member node data (may be a cluster node or a Pacemaker Remote node)
+typedef struct pcmk__cluster_member_s {
+    char *uname;                // Node name as known to cluster
+
+    /* @COMPAT This is less than ideal since the value is not a valid XML ID
+     * (for Corosync, it's the string equivalent of the node's numeric node ID,
+     * but XML IDs can't start with a number) and the three elements should have
+     * different IDs.
+     *
+     * Ideally, we would use something like node-NODEID, node_state-NODEID, and
+     * transient_attributes-NODEID as the element IDs. Unfortunately changing it
+     * would be impractical due to backward compatibility; older nodes in a
+     * rolling upgrade will always write and expect the value in the old format.
+     *
+     * This is also named poorly, since the value is not a UUID, but at least
+     * that can be changed at an API compatibility break.
+     */
+    /*! Value of the PCMK_XA_ID XML attribute to use with the node's
+     * PCMK_XE_NODE, PCMK_XE_NODE_STATE, and PCMK_XE_TRANSIENT_ATTRIBUTES
+     * XML elements in the CIB
+     */
+    char *uuid;
+
+    char *state;                // @TODO change to enum
+    uint64_t flags;             // Bitmask of crm_node_flags
+    uint64_t last_seen;         // Only needed by cluster nodes
+    uint32_t processes;         // @TODO most not needed, merge into flags
+
+    /* @TODO When we can break public API compatibility, we can make the rest of
+     * these members separate structs and use void *cluster_data and
+     * void *user_data here instead, to abstract the cluster layer further.
+     */
+
+    // Currently only needed by corosync stack
+    uint32_t id;                // Node ID
+    time_t when_lost;           // When CPG membership was last lost
+
+    // Only used by controller
+    enum crm_join_phase join;
+    char *expected;
+
+    time_t peer_lost;
+    char *conn_host;
+
+    time_t when_member;         // Since when node has been a cluster member
+    time_t when_online;         // Since when peer has been online in CPG
+} pcmk__cluster_member_t;
```

Comment:
This doesn't work for remote nodes either ... cluster member refers to being joined at the cluster layer. Peer refers to another member of the same CPG.

Maybe `pcmk__node_status_t`, since we're tracking the status of a live node vs storing basic information about a node.

Probably worthwhile making the type alias public; there's a good chance we'll expose some APIs with it eventually, and it'll avoid another rename.

Reply:
> cluster member refers to being joined at the cluster layer.

Yeah, I was hesitant for that reason. I felt like we were overloading "member" to include remote nodes in some cases, based on `cluster/membership.c`'s dealings with remote nodes. In other cases we do use it to mean joined cluster nodes, which seems more correct.

I think `pcmk__node_status_t` is pretty good and may be the best option. Only reluctance is that it sounds more related to `pcmk_node_t` than it actually is. We currently have `struct pcmk__scored_node`/`pcmk_node_t`, which is a wrapper for `struct pcmk__node_assignment` and the *real* node data `pcmk__node_private_t`. `pcmk__node_status_t` (currently `crm_node_t` is not directly connected to any of that.

Put differently, `pcmk__node_status_t` feels like it would be on a similar level or have a similar role to `struct pcmk__node_assignment`.

---

> Probably worthwhile making the type alias public; there's a good chance we'll expose some APIs with it eventually, and it'll avoid another rename.

Eh, we *can*, but I don't think a rename that simply drops one underscore is a big deal. `sed` it and forget it.